File: Leetcode/Non-overlapping Intervals/Non_overlapping_intervals.cpp
// * Algorithm ==>
// * Approach used : Longest Increasing Subsequence
// * In this we first sort the intervals array on the basis of first value. 
// * Then we calculate Longest Increasing Subsequence on the intervals array on the basis of the second value keeping in mind that south of the current element must be greater than north of the previous element. This is done in order to get non overlapping intervals.  
// * At last in order to get minimum intervals that can be removed , we subtract maximum non overlapping intervals from the length of intervals array, i.e. total - num of max non overlapping = min to be removed!
// * This approach uses O(n^2) 

class Solution
{
public:
    int eraseOverlapIntervals(vector<vector<int>> &intervals)
    {
        if (intervals.size() == 0)
        {
            return 0;
        }
        int n = (int)intervals.size(), _max = 1;
        sort(intervals.begin(), intervals.end());
        vector<int> dp(n, 1);
        for (int i = 1; i < n; i++)
        {
            for (int j = 0; j < i; j++)
            {
                if (intervals[j][1] <= intervals[i][0])
                {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
            _max = max(_max, dp[i]);
        }
        return n - _max;
    }
};

File: Leetcode/Non-overlapping Intervals/NonOverlappingIntervals.java
class Solution {
    public int maxEnvelopes(int[][] envelopes) {
        //sorting the intervals array based on first value
        Arrays.sort(envelopes, (a, b) -> Integer.compare(a[0], b[0]));

        int[] dp = new int[envelopes.length];
        Arrays.fill(dp, 1);

        // calculating LIS on the basis of second value in order to cover maximum no of intervals
        int max_=1;
        for(int i=1; i<envelopes.length; i++){
            for(int j=i-1; j>=0; j--){

                //including intervals to our answer only when we get start time of current < ending time of previous in order to avoid intersection
                if(envelopes[j][0] < envelopes[i][0] && envelopes[j][1]< envelopes[i][1]){
                    dp[i]= Math.max(dp[i], dp[j]+1);
                }
            }

            //calculating maxlength of dp .
            max_= Math.max(max_, dp[i]);
        }
        return max_;
    }
}